### PR TITLE
Remove a dead assignment of accChargeForToA in the HGCDigitizer

### PR DESCRIPTION
**Pull request by perrotta**
#### PR description:

Spotted by the static analyzer

#### PR validation:

It builds succesfully
No changes at all expected in the outputs

### CODE REVIEW

**Comment by perrotta** [comment]
assign hgcal-dpg

**Comment by perrotta** [comment]
please test

File: SimCalorimetry/HGCalSimProducers/plugins/HGCDigitizer.cc [hunk 1]
```diff
@@ -201,25 +200,25 @@ namespace {
...
+        float fireTDC(0.f);
         const auto aboveThrPos = std::upper_bound(
             hitRec.begin(), hitRec.end(), std::make_pair(cell_threshold, 0.f), [](const auto& i, const auto& j) {
               return i.first < j.first;
             });
 
         if (aboveThrPos == hitRec.end()) {
-          accChargeForToA = hitRec.back().first;
-          fireTDC = 0.f;
+          //          fireTDC = 0.f;
```
Comment (rovere) [inline SimCalorimetry/HGCalSimProducers/plugins/HGCDigitizer.cc:210]: Would it make the code more readable if the 2 conditions are combined?

Comment (perrotta) [reply SimCalorimetry/HGCalSimProducers/plugins/HGCDigitizer.cc:210]: > Would it make the code more readable if the 2 conditions are combined?

Why not....
It is up to you, anyhow. I did not want to modify too much the structure of the code, but if you want I can implement it here as well.

Comment (rovere) [reply SimCalorimetry/HGCalSimProducers/plugins/HGCDigitizer.cc:210]: 👍🏻

Comment (perrotta) [reply SimCalorimetry/HGCalSimProducers/plugins/HGCDigitizer.cc]: @rovere done

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
please test

**Comment by rovere** [comment]
Ciao @perrotta thanks for the changes. I see that all the failed comparisons belong to the only workflow using the `premixing` which is the only part of the code that this PR touches.
I'd have expected 0 regression unless there's some other randomization happening that spoils the reproducibility for different reasons.
Suggestinos?

**Comment by perrotta** [comment]
Indeed, there was a mistake...

**Comment by perrotta** [comment]
please test

**Comment by rovere** [comment]
+1

**Comment by civanch** [comment]
+1

File: SimCalorimetry/HGCalSimProducers/plugins/HGCDigitizer.cc [hunk 2]
```diff
@@ -201,25 +200,24 @@ namespace {
...
-          fireTDC = aboveThrPos->second;
-          if (aboveThrPos - hitRec.begin() >= 1) {
-            const auto& belowThrPos = aboveThrPos - 1;
-            float chargeBeforeThr = belowThrPos->first;
-            float timeBeforeThr = belowThrPos->second;
-            float deltaQ = accChargeForToA - chargeBeforeThr;
-            float deltaTOF = fireTDC - timeBeforeThr;
-            fireTDC = (cell_threshold - chargeBeforeThr) * deltaTOF / deltaQ + timeBeforeThr;
+        if (aboveThrPos != hitRec.end()) {
+          if (hitRec.end() - aboveThrPos > 0 || orderChanged) {
```
Comment (perrotta) [inline SimCalorimetry/HGCalSimProducers/plugins/HGCDigitizer.cc:210]: @rovere, if I am not wrong `aboveThrPos` can never be larger than `hitRec.end()`, it being bound between `hitRec.begin()` and `hitRec.end()`, see https://github.com/cms-sw/cmssw/pull/39622/files#diff-563eece7ae99b29c2501955e163a955bd6940d6fa39094e919eee33975bdb1f2R204-R207

Therefore, given the [L209](https://github.com/cms-sw/cmssw/pull/39622/files#diff-563eece7ae99b29c2501955e163a955bd6940d6fa39094e919eee33975bdb1f2R209) the check of `orderChanged` can never be done.

Please notice that this was the same already in the original code: my fix simply reproduces the original logic but avoiding the dead assignment. If this observation of mine is correct, the original code was already bugged, and the dead assignment reported by the static analyzer helped to spot the bug. (And if so, of course, this PR has to be updated with a more proper fix that I'd suggest either you or some other HGCal code developer take care of).

**Comment by srimanob** [comment]
+Upgrade

**Comment by srimanob** [comment]
@cmsbuild please test

Re-test after 3 weeks.

**Comment by perrotta** [comment]
@rovere @felicepantaleo , while tests are running could you please have a look at my old comment https://github.com/cms-sw/cmssw/pull/39622#discussion_r990763607 ?

**Comment by rappoccio** [comment]
hold

* Reminder that there is a pending question [here](https://github.com/cms-sw/cmssw/pull/39622#discussion_r990763607)

**Comment by rappoccio** [comment]
hold

**Comment by perrotta** [comment]
No need to keep this PR in standby forever only because it points out a bug in the code: an issue for it has been opened in https://github.com/cms-sw/cmssw/issues/40576 for that
Let rerun the tests here, and merge this technical PR if they still succed after 5 months

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
unhold

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1

**Pull request merged**